docs-src/design-decisions.md

```diff
@@ -0,0 +1,42 @@
+# Design Decisions
+
+Throughout the design of these APIs some significant design decisions have been
+made. This provides some background on the factors considered when making these
+decisions.
+
+## GatewayClass Scope
+
+### Problem
+There were a number of compelling reasons to make GatewayClass a
+namespace-scoped resource. 
+
+### Advantages of a Namespace-Scoped GatewayClass
+* Less potential for naming conflicts among different implementations.
+* Better modeling for implementations with multiple GatewayClass resources
+  (including some potential multi-tenant scenarios using Cluster API).
+* Easier to provide simple and secure defaults (only allow Gateways and Routes
+  in the same namespace by default).
+* GatewayClass parameters can live in the same namespace.
+* More RBAC flexibility (can allow users to create GatewayClasses in select
+  namespaces only or restrict visibility of GatewayClasses in some namespaces).
+
+### Advantages of a Cluster-Scoped GatewayClass
+* Consistent with StorageClass and IngressClass resources.
+* Discussions with sig-storage showed that StorageClass had worked well as a
+  cluster-scoped resource and that they were continuing to follow that pattern
+  for new Class resources.
+* Follows general pattern that resources managed by infrastructure provider
+  and/or cluster administrators are cluster scoped.
+* Impossible for namespace level admins to grant RBAC permissions for
+  GatewayClass resources. (With a namespace-scoped resource, anyone that could
+  configure RBAC within a namespace could grant GatewayClass access).
+* Clear that GatewayClasses can be used by multiple namespaces.
```

Comment:
Well, not really. Not with the current implementation that defines the `AllowedGatewayNamespaces` and `AllowedRouteNamespaces` on the `GatewayClass`.

Reply:
Yeah both a namespace-scoped and cluster-scoped GatewayClass would still be referenced by Gateways in other namespaces, it just would seem at least slightly less obvious that a namespace-scoped resource could be referenced by Gateways in other namespaces. This would go against patterns elsewhere in Kubernetes. For example, with RBAC, a Role can only be referenced by a RoleBinding in the same namespace. Similar models for Pods/Deployments/DaemonSets/etc that are inherently limited to a single namespace.

Reply:
I mean, if you keep `GatewayClass` as a cluster scoped resource then IMO the `Allowed` fields should be removed. If you want that, then make GWC a namespace scoped resource or create  new API that is.

Reply:
Unfortunately even a namespace-scoped resource wouldn't solve this. We definitely want it to be possible to reference a GatewayClass from multiple namespaces. A namespace-scoped resource would make it straightforward to allow only-local references of a GatewayClass, but I don't think it helps with the very real use case of "only Gateways in namespaces labeled `env: prod` should be able to use this GatewayClass".

Reply:
That’s fair, but it’s one thing to define that on a namespaced type and another on a cluster scoped resource that is supposed to satisfy all comers.

Reply:
Absent a general form of hierarchy, we don't have a way to express who can see a thing other than "don't cross namespaces.

There are a LOT of things I would change about how k8s does this stuff, if I could do so with a reasonable amount of energy.  TTBOMK "who can use" has not come up for StorageClass or RuntimeClass yet.